pkg/distributor/influxpush/parser.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	encoding := r.Header.Get("Content-Encoding")
+	reader, err := batchReadCloser(r.Body, encoding, int64(maxSize))
+	if err != nil {
+		return nil, 0, fmt.Errorf("gzip compression error: %w", err)
+	}
+	data, err := io.ReadAll(reader)
+	dataLen := len(data) // In case if something is read despite an error.
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't read body: %s", err)
+	}
+
+	err = reader.Close()
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("problem reading body: %s", err)
+	}
+
+	points, err := models.ParsePointsWithPrecision(data, time.Now().UTC(), precision)
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't parse points: %s", err)
+	}
+	ts, err := writeRequestFromInfluxPoints(points)
+	return ts, dataLen, err
+}
+
+func writeRequestFromInfluxPoints(points []models.Point) ([]mimirpb.PreallocTimeseries, error) {
+	// Technically the same series should not be repeated. We should put all the samples for
+	// a series in single client.Timeseries. Having said that doing it is not very optimal and the
+	// occurrence of multiple timestamps for the same series is rare. Only reason I see it happening is
+	// for backfilling and this is not the API for that. Keeping that in mind, we are going to create a new
+	// client.Timeseries for each sample.
+
+	returnTs := mimirpb.PreallocTimeseriesSliceFromPool()[:0]
+	if cap(returnTs) < len(points) {
+		returnTs = make([]mimirpb.PreallocTimeseries, 0, len(points))
+	}
+	for _, pt := range points {
+		var err error
+		returnTs, err = influxPointToTimeseries(pt, returnTs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return returnTs, nil
+}
+
+// Points to Prometheus is heavily inspired from https://github.com/prometheus/influxdb_exporter/blob/a1dc16ad596a990d8854545ea39a57a99a3c7c43/main.go#L148-L211
+func influxPointToTimeseries(pt models.Point, returnTs []mimirpb.PreallocTimeseries) ([]mimirpb.PreallocTimeseries, error) {
+
+	fields, err := pt.Fields()
+	if err != nil {
+		return returnTs, fmt.Errorf("can't get fields from point: %w", err)
+	}
+	for field, v := range fields {
+		var value float64
+		switch v := v.(type) {
+		case float64:
+			value = v
+		case int64:
+			value = float64(v)
+		case bool:
+			if v {
+				value = 1
+			} else {
+				value = 0
+			}
+		default:
+			continue
+		}
+
+		name := string(pt.Name()) + "_" + field
+		if field == "value" {
+			name = string(pt.Name())
+		}
+		replaceInvalidChars(&name)
+
+		tags := pt.Tags()
+		lbls := make([]mimirpb.LabelAdapter, 0, len(tags)+2) // An additional one for __name__, and one for internal label
+		lbls = append(lbls, mimirpb.LabelAdapter{
+			Name:  labels.MetricName,
+			Value: name,
+		})
+		lbls = append(lbls, mimirpb.LabelAdapter{
+			Name:  internalLabel, // An internal label for tracking active series
+			Value: "influx",
+		})
+		for _, tag := range tags {
+			key := string(tag.Key)
+			if key == "__name__" || key == internalLabel {
+				continue
+			}
+			replaceInvalidChars(&key)
```

Comment:
We create a string from the `tag.Key` slice (first allocation), then we pass it as a pointer to `replaceInvalidChars` which allocates a slice from that string (another allocation), and then we write into that pointer the string we allocate (third allocation) from that bytes slice.

Why don't we just do `key := replaceInvalidChars(tag.Key)`, and make `replaceInvalidChars` just modify the incoming bytes slice and then return it?

Reply:
Done as agreed...